### ISSUE TITLE
Proposal for Promises API

### ISSUE THREAD

**Issue by thanpolas** [jpodwys/cache-service#19]
By chance of supporting Promises API for all methods we need to have a deeper look at the `.set()` method which with a variable argument count and two functions (refresh and callback) can be problematic in implementing promises transparently...

The problem is that we will not be able to determine if the last arg of `.set()` is a callback or a refresh method and the user uses the Promises API.

```js
cs.set(key, value, refreshCb).then(function(){});
```

This will not work because the lib will assume that `refreshCb` is the actual callback thus using that for response vs the Promises API.

The only way I can see us going around that is to have all options for `set()` in a dictionary object:

```
```js
cs.set(key, value, {refresh: refreshCb}).then(function(){});

// using expiration too
cs.set(key, value, {
  refresh: refreshCb,
  expiration: 200,
}).then(function(){});

// Vanilla callback
cs.set(key, value, opts, callback);
```

Should we go with the `opts` solution we'll then need to have a look at `mget()` as well for API uniformity.

what do you think?

**Comment by jpodwys** [comment]
Some thoughts:

1. You've made excellent and thorough points. I like the dictionary object. 
2. I like to keep my cache modules in functional sync but I'm not sure how I feel about promisifying all of them natively, especially cache-service-cache-module since doing so could bloat the browser download. I think I would prefer to allow users to pass in their own promise library and take advantage of that so we don't force an increased download size for users that prefer callbacks. 
3. If possible, I'd prefer to keep backward compatibility and not make the dictionary object force a breaking change. Perhaps the dictionary object is active when promises are used or something.

**Comment by thanpolas** [comment]
>  especially cache-service-cache-module since doing so could bloat the browser download.

ah, this is used on the browser too? ok we can wrap everything in the cache-service module alone...

Re point 3 we can mangle it, if an object is passed it's easily detectable and we can have both APIs in place... We should set a best practice though, it won't be a breaking change but a deprecation notice... 

Current: 1.x
Promises API, non backwards breaking compat: 2.x (up next)
Promises API, backwards breaking (accepts only dict): 3.x (in a year or so)

**Comment by jpodwys** [comment]
This _can_ be used on the client, but it's more likely that `cache-service-cache-module` is used on the client on its own or as a dependency of `superagent-cache`. It's important to me that the APIs for all my cache modules, including `cache-service`, remain as identical as possible because I've promised they're swappable within both `cache-service` and `superagent-cache`.

**Comment by marybeshaw** [comment]
My team uses cache-service with the Redis module, and I thought I'd chime in for what it's worth.
I really like the dictionary object suggestion for the API, and would not have a problem with updating our code when it becomes available.

**Comment by jpodwys** [comment]
@marybeshaw thanks for your input!

**Comment by thanpolas** [comment]
new release is cool...

With proper semversioning we'll be cool for all the cache modules and `cache-service` as well as `superagent-cache`, we jump the major number on all the packages horizontally and we're good with API changes as far as contracts go...

I had a look at the `set()` definition in the redis cache lib and saw that there too is the `refresh` callback concept along with a background mechanism to interval it... Why is this necessary? Why can't `cache-service` and `superagent-cache` handle the background cache refresh and we have to include it in each caching module separately again and again?


So, bottom line here is that if we want the Promises API on the `cache-service` package we can't do it without slightly changing the API of `set()` and `mset()` to accept a dict of options. If we won't do that there will be dangerous ambiguity on `.set()` regarding the `refresh` function and the callback. So a decision has to be made for that.

**Comment by jpodwys** [comment]
All of the cache modules function on their own as complete entities. If I moved background refresh out of the individual modules it would make their APIs less robust. Additionally, cache-service and superagent-cache do not reimplement background refresh, they simply take advantage of the background refresh feature made available by whichever cache module is used.

**Comment by JaredReisinger** [comment]
FWIW, I don't have any problems using promises with cache-service and the modules.  I can only speak to Bluebird's implementation of promisify, but all it does is take the arguments you pass to the generated wrapper, append the promisifying callback, and call the underlying implementation.  From cache-service's point of view, this looks exactly like a regular, non-promisifed call.  More to the point, cache-service specifically documents that the last function is _always_ the callback, as per the Node standard.

Okay, I fibbed a little.  There is a problem of sorts, at least with my expectations for the timing of completion, and I've raised it as a separate issue (https://github.com/jpodwys/cache-service/issues/22).  For `.set()` (and `.mset()`, `.del()`, and `.flush()`), the provided callback—whether a promisified one or not—is called when the _first_ cache finishes, rather than when they _all_ finish.
